pkg/compactor/shard_aware_deduplicate_filter.go

```diff
@@ -200,24 +202,31 @@ func (b *blockWithSuccessors) isIncludedIn(other *blockWithSuccessors) bool {
 	return true
 }
 
-// addSuccessorIfPossible adds successor to this block, but only if this block is included in the other block.
-// Successor is not added directly to this block, but correct place in the tree is found for it first.
+// addSuccessorIfPossible adds the given block as a successor of this block, or of a successor of this block
+// Successor is added in the correct place in the tree of successors of this block.
 // Returns true, if other block was added as successor (somewhere in the tree), false otherwise.
-//
-// Since different subtrees overlap, we need to handle the case when this block and other block are the same pointer.
 func (b *blockWithSuccessors) addSuccessorIfPossible(other *blockWithSuccessors) bool {
+	// if this and other block are the same pointer,
+	// or if this block is not included in the other block, false is returned
 	if b == other || !b.isIncludedIn(other) {
 		return false
 	}
 
-	// Try to find ALL successors where we can include the block
+	// if the other block is already in the list of direct successors of this block, true is returned
+	if _, ok := b.successors[other.meta.ULID]; ok {
+		return true
+	}
```

Comment:
I don't think this can be considered a refactoring, since it's a behavioral change (I don't know the effect, I should say)?

Reply:
It's an optimization from what I can tell. It should be safe.

Reply:
@pstibrany thanks for checking :) I think it's misleading wrt. the PR declaring a straight refactoring is all.

Reply:
@aknuds1 I have changed the commit message and title to "Small code and documentation refactoring and improvements"